lib/src/kbasesearchengine/main/IndexerWorker.java

```diff
@@ -213,25 +213,41 @@ private void logError(
         logError(msg, e);
     }
     
-    private boolean performOneTick() throws InterruptedException, IndexingException {
-        final Optional<StoredStatusEvent> optEvent = retrier.retryFunc(
-                s -> s.setAndGetProcessingState(StatusEventProcessingState.READY, workerCodes,
-                        StatusEventProcessingState.PROC, id),
-                storage, null);
+    /** Runs one cycle of the event processing loop, processing up to one event.
+     * @return true if an event was processed, false if not.
+     * @throws InterruptedException if the thread was interrupted.
+     * @throws FatalIndexingException if an indexing exception occurred that should cause the
+     * shutdown of the worker. In normal use, no more events will be processed.
+     */
+    public boolean runCycle() throws InterruptedException, FatalIndexingException {
+        final Optional<StoredStatusEvent> optEvent;
+        try {
+            optEvent = retrier.retryFunc(
+                    s -> s.setAndGetProcessingState(StatusEventProcessingState.READY, workerCodes,
+                            StatusEventProcessingState.PROC, id),
+                    storage, null);
+        } catch (FatalIndexingException e) {
+            throw e;
+        } catch (IndexingException e) { // untestable
+            throw new RuntimeException("non-fatal exceptions should not be thrown here");
+        }
         boolean processedEvent = false;
         if (optEvent.isPresent()) {
             final StoredStatusEvent parentEvent = optEvent.get();
             final EventHandler handler;
             try {
                 handler = getEventHandler(parentEvent);
             } catch (UnprocessableEventIndexingException e) {
-                logError(LogPrefix.STD, e);
-                markEventProcessed(parentEvent, StatusEventProcessingState.FAIL);
+                handleException("Error getting event handler", parentEvent, e);
                 return true;
```

Comment:
Shouldn't this line return false?

Reply:
No, the boolean tells the caller whether an event was processed or not, regardless of the final state. The purpose of the boolean is so the calling loop can continue to pull events until there are none left with no delay between events. When this function returns false, the loop will wait.